### PR TITLE
introduce peer manager

### PR DESCRIPTION
**Pull request by Liuhaai**
### Description 
 - The new component `peerManager` is introduced for peer discovering, peer connecting, and peer managing. In the `peerManager`, the [Rendezvous protocol](https://github.com/libp2p/go-libp2p/tree/master/examples/chat-with-rendezvous) is used for peer discovery, which replaces the old `Neighbors` method

 - To interact with `peerManager` externally, the methods `Advertise`, `FindPeers`, `BlockPeer`, `AddBootstrap` and `ConnectedPeers` are added in `Host` class.

 - Code clean work for `NewHost` func

### CODE REVIEW

File: peerManager.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	return err
+}
+
+func (ma *peerManager) ConnectPeers(ctx context.Context) error {
+	// skip connecting when 80% peers are connected
+	if len(ma.host.Network().Conns()) > ma.maxPeers*8/10 {
+		return nil
+	}
+	if ma.routing == nil {
+		return errNoDiscoverer
```
Comment (dustinxie) [inline peerManager.go:51]: use panic? b/c if `routing == nil`, the entire thing is useless

File: p2p.go [hunk 2]
```diff
@@ -380,9 +402,36 @@ func NewHost(ctx context.Context, options ...Option) (*Host, error) {
...
+// ConnectPeers connect host to the peers with the same groupID
+func (h *Host) ConnectPeers(ctx context.Context) error {
+	if h.peerManager == nil {
+		return errNoPeerManager
+	}
+	if err := h.peerManager.ConnectPeers(ctx); err != nil {
+		return err
+	}
+	// setup a goroutine by caller?
+	// TODO: stop goroutine when p2p stop()
```
Comment (dustinxie) [inline p2p.go:420]: yeah, i think this go-routine can be done in core's p2p using a `RecurringTask`?

File: p2p.go [hunk 3]
```diff
@@ -100,6 +103,8 @@ var (
 		RateLimit:                DefaultRatelimitConfig,
 		PrivateNetworkPSK:        "",
 		ProtocolID:               "/iotex",
+		PeersRefreshInterval:     30 * time.Minute,
+		GroupID:                  "iotex",
```
Comment (dustinxie) [inline p2p.go:107]: we can use `iotex + strconv.Itoa(int(chainID)` similar to ProtocolID(), then don't need this value in config

File: peerManager.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		pm.maxPeers = num
+	}
+}
+
+func withBlacklistTolerance(num int) peerManagerOpt {
+	return func(pm *peerManager) {
+		pm.blacklistTolerance = num
+	}
+}
+
```
Comment (dustinxie) [inline peerManager.go:39]: add a `func withBlackListTimeout(time.Duration) opt`

and add a timeout value into config

File: p2p.go [hunk 5]
```diff
@@ -66,6 +65,8 @@ type (
 		EnableRateLimit          bool            `yaml:"enableRateLimit"`
 		RateLimit                RateLimitConfig `yaml:"rateLimit"`
 		PrivateNetworkPSK        string          `yaml:"privateNetworkPSK"`
+		GroupID                  string          `yaml:"groupID"`
+		ServerMode               bool            `yaml:"serverMode"`
```
Comment (dustinxie) [inline p2p.go:69]: 1. remove `ServerMode`? we discussed it needs to be server mode, so no need to change regarding this
2. add a blacklist timeout value, see comment below

File: p2p.go [hunk 6]
```diff
@@ -178,6 +182,14 @@ func MasterKey(masterKey string) Option {
...
 }
 
+// ServerMode is turn node into p2p bootstrap
+func ServerMode() Option {
+	return func(cfg *Config) error {
+		cfg.ServerMode = true
+		return nil
+	}
+}
+
```
Comment (dustinxie) [inline p2p.go:192]: remove

File: p2p.go [hunk 7]
```diff
@@ -331,7 +354,13 @@ func NewHost(ctx context.Context, options ...Option) (*Host, error) {
 	if err != nil {
 		return nil, err
 	}
-	kad, err := dht.New(ctx, host, dht.ProtocolPrefix(cfg.ProtocolID), dht.Mode(dht.ModeServer))
+	dhtOpt := []dht.Option{dht.ProtocolPrefix(cfg.ProtocolID)}
+	if cfg.ServerMode {
+		dhtOpt = append(dhtOpt, dht.Mode(dht.ModeServer))
+	} else {
+		dhtOpt = append(dhtOpt, dht.Mode(dht.ModeClient))
+	}
```
Comment (dustinxie) [inline p2p.go:362]: remove

File: p2p.go [hunk 8]
```diff
@@ -380,9 +408,22 @@ func NewHost(ctx context.Context, options ...Option) (*Host, error) {
 }
 
 // JoinOverlay triggers the host to join the DHT overlay
-func (h *Host) JoinOverlay(ctx context.Context) {
+func (h *Host) JoinOverlay(ctx context.Context) error {
 	routingDiscovery := discovery.NewRoutingDiscovery(h.kad)
-	discovery.Advertise(ctx, routingDiscovery, h.kadKey.String())
+	h.peerManager = newPeerManager(h.host, routingDiscovery, h.cfg.GroupID, withMaxPeers(100), withBlacklistTolerance(3))
```
Comment (dustinxie) [inline p2p.go:413]: move to `h.New()`

File: p2p.go [hunk 9]
```diff
@@ -380,9 +408,22 @@ func NewHost(ctx context.Context, options ...Option) (*Host, error) {
...
+	return h.peerManager.Advertise(ctx)
+}
+
+// TODO: stop goroutine when p2p stop()
+// FindPeers connect host to the peers with the same groupID
+func (h *Host) FindPeers(ctx context.Context) error {
+	if h.peerManager == nil {
+		return errNoPeerManager
+	}
```
Comment (dustinxie) [inline p2p.go:422]: can remove this check and `errNoPeerManager`? `h.peerManager` will be created inside `New()`

File: p2p.go [hunk 10]
```diff
@@ -542,36 +573,49 @@ func (h *Host) Broadcast(ctx context.Context, topic string, data []byte) error {
...
 }
 
 // ClearBlocklist clears the blocklist
 func (h *Host) ClearBlocklist() {
-	h.unicastBlocklist.Clear()
+	h.peerManager.ClearBlockList()
+}
+
+// AddBlocklist clears the blocklist
+func (h *Host) AddBlocklist(id peer.ID) {
```
Comment (dustinxie) [inline p2p.go:605]: remove this func? it is not used

File: p2p.go [hunk 11]
```diff
@@ -542,36 +573,49 @@ func (h *Host) Broadcast(ctx context.Context, topic string, data []byte) error {
...
+	h.peerManager.BlockPeer(id)
+}
+
+// AddBootstrap adds bootnode into peermanager
+func (h *Host) AddBootstrap(bootNodeAddrs []multiaddr.Multiaddr) error {
```
Comment (dustinxie) [inline p2p.go:610]: remove this func? it is not used

File: peerManager.go [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+	}
+	return pm
+}
+
+func (ma *peerManager) Advertise(ctx context.Context) error {
+	_, err := ma.routing.Advertise(ctx, ma.ns)
+	return err
+}
+
+func (ma *peerManager) AddBootstrap(ids ...core.PeerID) {
```
Comment (dustinxie) [inline peerManager.go:81]: remove this func? it is not used

File: p2p.go [hunk 13]
```diff
@@ -21,17 +20,17 @@ import (
...
 	"github.com/libp2p/go-libp2p-core/pnet"
 	"github.com/libp2p/go-libp2p-core/protocol"
-	"github.com/libp2p/go-libp2p-discovery"
-	"github.com/libp2p/go-libp2p-kad-dht"
-	"github.com/libp2p/go-libp2p-pubsub"
-	"github.com/libp2p/go-libp2p-secio"
+	discovery "github.com/libp2p/go-libp2p-discovery"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	secio "github.com/libp2p/go-libp2p-secio"
```
Comment (dustinxie) [inline p2p.go:33]: can this revert?

File: peerManager.go [hunk 14]
```diff
@@ -157,6 +157,11 @@ func (ma *peerManager) BlockPeer(pr peer.ID) {
 	}
 }
 
+func (ma *peerManager) BlockPeer(pr peer.ID) {
```
Comment (dustinxie) [inline peerManager.go:160]: call this `RemovePeer`?

**Review by dustinxie — APPROVED** [review]

File: peerManager.go [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+func newPeerManager(host core.Host, routing *discovery.RoutingDiscovery, ns string, opts ...peerManagerOpt) *peerManager {
+	rand.Seed(time.Now().UnixNano())
+	pm := &peerManager{
+		host:              host,
+		routing:           routing,
+		ns:                ns,
+		bootstrap:         make(map[peer.ID]bool),
+		advertiseInterval: 3 * time.Hour, // DHT provider record is republished every 3hrs by default
```
Comment (Liuhaai) [inline peerManager.go:72]: The group id will expire in 24 hrs. So it's advertised forcely every 3 hrs.

File: p2p.go [hunk 16]
```diff
@@ -417,8 +452,19 @@ func (h *Host) AddUnicastPubSub(topic string, callback HandleUnicast) error {
 			Logger().Error("Error when subscribing a unicast message.", zap.Error(err))
 			return
 		}
-		ctx := context.WithValue(context.Background(), unicastCtxKey{}, stream)
-		if err := callback(ctx, stream, data); err != nil {
+
+		conn := stream.Conn()
+		remoteID := conn.RemotePeer()
+		RemoteAddr := peer.AddrInfo{
+			ID:    remoteID,
```
Comment (dustinxie) [inline p2p.go:459]: `ID: conn.RemotePeer()` so we don't need line 457

File: peerManager.go [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+		opt(pm)
+	}
+	if pm.blacklistTimeout > 0 {
+		bl, _ := ttl.NewCache(ttl.AutoExpireOption(pm.blacklistTimeout))
+		pm.blacklist = bl
+	} else {
+		bl, _ := ttl.NewCache()
+		pm.blacklist = bl
+	}
+
```
Comment (dustinxie) [inline peerManager.go:84]: remove empty line

File: peerManager.go [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+				_, err := pm.routing.Advertise(context.Background(), pm.ns)
+				if err != nil {
+					Logger().Error("error when advertising.", zap.Error(err))
+				}
+			}
+		}
+	}(pm)
+}
+
+func (pm *peerManager) Advertise() error {
```
Comment (dustinxie) [inline peerManager.go:109]: also we don't need this? in `JoinOverlay()`, `ticker.C` will fire every 3 hours to keep advertising itself, right? maybe we can reduce to interval to like 30 min

File: peerManager.go [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+		}
+	}
+	return false
+}
+
+func (pm *peerManager) DisconnectPeer(pr peer.ID) error {
+	return pm.host.Network().ClosePeer(pr)
+}
+
+func (pm *peerManager) TryBlockPeer(pr peer.ID) {
```
Comment (dustinxie) [inline peerManager.go:184]: this name is not accurate

File: peerManager.go (hunk 18 above)
Comment (Liuhaai) [reply peerManager.go:109]: this is the func triggerred externally

File: peerManager.go (hunk 19 above)
Comment (Liuhaai) [reply peerManager.go:184]: Suggestion?

File: peerManager.go [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+	val, exist := pm.blacklist.Get(pr)
+	if !exist {
+		pm.blacklist.Set(pr, 1)
+		return
+	}
+	pm.blacklist.Set(pr, val.(int)+1)
+	if val.(int)+1 >= pm.blacklistTolerance {
+		pm.host.Network().ClosePeer(pr)
+	}
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply peerManager.go:193]: The peer will be filtered out by `blocked()`, so the node can't get this peer by `ConnectedPeers()` . And the id will be removed from the map when BlacklistTimeout reaches.

File: peerManager.go [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+		pm.blacklist = bl
+	} else {
+		bl, _ := ttl.NewCache()
+		pm.blacklist = bl
+	}
+
+	return pm
+}
+
+func (pm *peerManager) JoinOverlay() {
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply peerManager.go:88]: Good suggestion

**Pull request merged**